monitoring/queuemon_test.go

```diff
@@ -0,0 +1,186 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package monitoring
+
+import (
+	"encoding/json"
+	"io"
+	"net/http"
+	"runtime"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/opt"
+	"github.com/elastic/beats/v7/libbeat/publisher/queue"
+	"github.com/elastic/elastic-agent-shipper/monitoring/reporter/expvar"
+)
+
+// emulates what the expvar queue metrics look like on the other end
+
+type expvarQueue struct {
+	CurrentLevel int  `json:"current_queue_level"`
+	Maxlevel     int  `json:"max_queue_level"`
+	IsFull       bool `json:"queue_is_currently_full"`
+	LimitCount   int  `json:"queue_limit_reached_count"`
+}
+
+// ======= mocked test input queue
+
+// TestMetricsQueue is a test queue for the reporter
+type TestMetricsQueue struct {
+	metricState queue.Metrics
+	limit       uint64
+}
+
+// NewTestQueue returns a new test queue
+func NewTestQueue(limit uint64) *TestMetricsQueue {
+	return &TestMetricsQueue{
+		metricState: queue.Metrics{
+			OldestActiveTimestamp: common.Time(time.Now()),
+			EventCount:            opt.UintWith(0),
+			EventLimit:            opt.UintWith(limit),
+		},
+		limit: limit,
+	}
+}
+
+// BufferConfig doesn't do anything
+func (tq TestMetricsQueue) BufferConfig() queue.BufferConfig {
+	return queue.BufferConfig{}
+}
+
+// Producer doesn't do anything
+func (tq TestMetricsQueue) Producer(_ queue.ProducerConfig) queue.Producer {
+	return nil
+}
+
+// Consumer doesn't do anything
+func (tq TestMetricsQueue) Consumer() queue.Consumer {
+	return nil
+}
+
+// Close Doesn't do anything
+func (tq TestMetricsQueue) Close() error {
+	return nil
+}
+
+// Metrics spoofs the metrics output
+func (tq *TestMetricsQueue) Metrics() (queue.Metrics, error) {
+	//fmt.Printf("Got queue state: %#v\n", tq)
+	tq.metricState.EventCount = opt.UintWith(tq.metricState.EventCount.ValueOr(0) + 1)
+
+	if tq.metricState.EventCount.ValueOr(0) > tq.limit {
+		tq.metricState.EventCount = opt.UintWith(0)
+	}
+
+	return tq.metricState, nil
+}
+
+// ============ mocked output reporter
+
+// simple wrapper to return a generic config object
+func initMonWithconfig(interval int, name string) Config {
+	return Config{
+		Interval: time.Millisecond * time.Duration(interval),
+		Enabled:  true,
+		ExpvarOutput: expvar.Config{
+			Enabled: true,
+			Port:    8081,
```

Comment:
I think this is always starting a server listening on localhost port 8081. What you probably want to do instead if listen on a randomly chosen unallocated port (port 0 can request this). That will prevent port collisions in tests.

You can use https://pkg.go.dev/net/http/httptest#NewServer to wrap this up in a new way but you'd have to add a way to change what method is used to start the server.

Reply:
Yah, agreed, that might be what is nuking the tests, since it works on my system.